Part3 project setup

hiandra opened the pull request.

hiandra committed:
[Setup of Part 3]

   - Added files for the part 3 of the project.
   - A new directory called reachability was created.

DavidFosca committed:
[Start - Implementation of Reachability]

   - Implementation of constructor.

hiandra committed:
[Implementation of ReachabilityInterface Functions]

   - Implemented the interface and auxiliary function.
   - Created tests for the interface based on the example
     provided in the instructions for part 3.

hiandra committed:
[Added reachability to GitHub Actions]

   - Edited the actions file to add reachability tests.

DavidFosca commented on src/reachability/Reachability.cpp:
```diff
@@ -0,0 +1,164 @@
+    BDD_ID bddIDVariable;
+
+    if (stateSize==0) {
+        throw std::runtime_error("Runtime_error if stateSize is zero!");
```
We still have to check if the runtime_errors are working ok

DavidFosca commented on src/reachability/Reachability.cpp:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    for(int i=0; i<transitionFunctions.size(); i++){
+        if (transitionFunctions[i] > uniqueTableSize()-2) {
```
Why Size-2? (just to understand)

hiandra committed:
[Fixed bugs in isReachable]

   - Added throw runtime error in isReachable.
   - Fixed case when isReachable is called without
     calling the setTransitionFunctions before.
   - Added test for 2 bits when using the default
     transition function.
   - Added test for 3 bits - fails.

hiandra committed:
[Fixed bugs in the case of different state bits]

   - Setting the transitionFunctions was missing when
     the initial states were changed (set).
   - Added tests for all the exceptions.
   - Tests: passed.

DavidFosca reviewed (approved):
ok:)
